benchmarks/experiment_runner.py

```diff
@@ -183,22 +188,145 @@ def run_single_config(self):
 
+    # Repeat the experiment and accumulate metrics.
     with benchmark_model.pick_grad():
-      metrics = OrderedDict()
+      accumulated_metrics = OrderedDict()
       for repeat_iteration in range(self._args.repeat):
-        run_metrics = self.timed_run(benchmark_experiment, benchmark_model,
-                                     experiment_config, model_config,
-                                     repeat_iteration)
-
-        for key, val in run_metrics.items():
-          if key not in metrics:
-            metrics[key] = []
-          metrics[key].append(val)
+        metrics = self.run_once_and_gather_metrics(benchmark_experiment,
+                                                   benchmark_model,
+                                                   experiment_config,
+                                                   model_config,
+                                                   repeat_iteration)
+        for k, v in metrics.items():
+          if k not in accumulated_metrics:
+            accumulated_metrics[k] = []
+          accumulated_metrics[k].append(v)
 
     # TODO: Use `experiment_config` and `model_config` when env vars are no
     # longer included.
     self._save_results(benchmark_experiment.to_dict(),
-                       benchmark_model.to_dict(), metrics)
+                       benchmark_model.to_dict(), accumulated_metrics)
+
+  def run_once_and_gather_metrics(self, benchmark_experiment, benchmark_model,
+                                  experiment_config, model_config,
+                                  repeat_iteration: int):
+
+    # Prepare inputs.
+    reset_rng_state(benchmark_experiment)
+    inputs_list = self._prepare_inputs(benchmark_model.example_inputs,
+                                       self._args.randomize_input)
+
+    # Reset state and sync.
+    reset_rng_state(benchmark_experiment)
+    self._mark_step(benchmark_experiment)
+    self._synchronize(benchmark_experiment)
+    met.clear_all()
+    metrics = OrderedDict()
+
+    # Start timers.
+    t_start = time.perf_counter()
+    if benchmark_experiment.xla:
+      t_trace = 0
+
+    def loop(pytorch_profile=None):
+      nonlocal t_trace
+      for i in range(self._args.iterations_per_run):
+
+        # Invoke iteration function and measure tracing time w/o waiting on the
+        # result.
+        if benchmark_experiment.xla:
+          t_trace_start = time.perf_counter()
+        output = benchmark_model.model_iter_fn(
+            inputs_list[i], collect_full_output=self._args.collect_full_output)
+        if benchmark_experiment.xla:
+          t_trace += time.perf_counter() - t_trace_start
+
+        # Mark step.
+        self._mark_step(benchmark_experiment)
+        if pytorch_profile is not None:
+          pytorch_profile.step()
+
+      self._synchronize(benchmark_experiment)
+      return output
+
+    # Execute all iterations (with) profiling.
+    enable_pytorch_profiling = self._args.dump_pytorch_profiles or \
```

Comment:
FYI previous semantics was that you had to run with `--profile-cuda`, and then choose one (or two) of both options: collecting numbers to the `results.jsonl` file, and/or dumping traces with kernel summaries. 

Probably we can get rid of the first one.

Reply:
Done.